File: uiframe/src/main/java/mobi/cangol/mobile/navigation/TabDrawerNavigationFragmentActivity.java
package mobi.cangol.mobile.navigation;

import android.app.Activity;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.drawerlayout.widget.DrawerLayout;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import mobi.cangol.mobile.base.BaseFragment;
import mobi.cangol.mobile.base.BaseNavigationFragmentActivity;
import mobi.cangol.mobile.uiframe.R;

/**
 * Created by xuewu.wei on 2017/9/19.
 */

public abstract class TabDrawerNavigationFragmentActivity extends BaseNavigationFragmentActivity {

    public static final String GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE = "getActivity is TabDrawerNavigationFragmentActivityDelegate";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        this.setNavigationFragmentActivityDelegate(new TabDrawerNavigationFragmentActivityDelegate(
                this));
        super.onCreate(savedInstanceState);
        this.getCustomActionBar().setTitleGravity(Gravity.CENTER);
        this.getCustomActionBar().setDisplayShowHomeEnabled(false);
        //默认是不能滑动的
        this.setDrawerEnable(Gravity.LEFT, false);
        this.setDrawerEnable(Gravity.RIGHT, false);
    }

    @Override
    public boolean onSupportNavigateUp() {
        if (stack.size() <= 1) {
            stack.peek().onSupportNavigateUp();
            return true;
        } else {
            return super.onSupportNavigateUp();
        }
    }

    public void setDrawerEnable(int gravity, boolean enable) {
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).setDrawerEnable(gravity, enable);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }

    public void showDrawer(int gravity, boolean show) {
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).showDrawer(gravity, show);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }

    public boolean isShowDrawer(int gravity) {
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            return ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).isShowDrawer(gravity);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }

    public void setDrawer(int gravity, Class<? extends BaseFragment> fragmentClass, Bundle args) {
        BaseFragment drawerFragment = (BaseFragment) Fragment.instantiate(this, fragmentClass.getName(), args);
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).setDrawer(gravity, drawerFragment);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }

    public BaseFragment getDrawer(int gravity) {
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            return ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).getDrawer(gravity);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }

    public void removeDrawer(int gravity) {
        if (getNavigationFragmentActivityDelegate() instanceof TabDrawerNavigationFragmentActivityDelegate) {
            ((TabDrawerNavigationFragmentActivityDelegate) getNavigationFragmentActivityDelegate()).removeDrawer(gravity);
        } else {
            throw new IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE);
        }
    }
}

class TabDrawerNavigationFragmentActivityDelegate extends AbstractNavigationFragmentActivityDelegate {

    private BaseNavigationFragmentActivity mActivity;
    private TabMenuDrawerLayout mDrawerLayout;
    private FrameLayout mMenuView;
    private FrameLayout mContentView;

    public TabDrawerNavigationFragmentActivityDelegate(
            BaseNavigationFragmentActivity activity) {
        mActivity = activity;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        mDrawerLayout = (TabMenuDrawerLayout) LayoutInflater.from(mActivity).inflate(
                R.layout.navigation_tab_drawer_main, null);

        mContentView = (FrameLayout) mDrawerLayout.findViewById(R.id.content_view);
        mMenuView = (FrameLayout) mDrawerLayout.findViewById(R.id.menu_view);
        mDrawerLayout.setDrawerListener(new DrawerLayout.DrawerListener() {
            @Override
            public void onDrawerSlide(View drawerView, float slideOffset) {
                Fragment fragment = mActivity.getSupportFragmentManager().findFragmentById(drawerView.getId());
                if (fragment instanceof BaseFragment) {
                    ((BaseFragment) fragment).onDrawerSlide(slideOffset);
                }
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                Fragment fragment = mActivity.getSupportFragmentManager().findFragmentById(drawerView.getId());
                if (fragment instanceof BaseFragment) {
                    fragment.setUserVisibleHint(true);
                    ((BaseFragment) fragment).onDrawerOpened();
                }
            }

            @Override
            public void onDrawerClosed(View drawerView) {
                Fragment fragment = mActivity.getSupportFragmentManager().findFragmentById(drawerView.getId());
                if (fragment instanceof BaseFragment) {
                    fragment.setUserVisibleHint(false);
                    ((BaseFragment) fragment).onDrawerClosed();
                }
            }

            @Override
            public void onDrawerStateChanged(int newState) {
                //do nothings
            }
        });
    }

    public void setDrawerEnable(int gravity, boolean enable) {
        mDrawerLayout.setDrawerEnable(gravity, enable);
    }
    public FrameLayout getMaskView() {
        return mDrawerLayout.getMaskView();
    }

    public void displayMaskView(boolean show) {
        mDrawerLayout.displayMaskView(show);
    }
    public void showDrawer(int gravity, boolean show) {
        mDrawerLayout.showDrawer(gravity, show);
    }

    public boolean isShowDrawer(int gravity) {
        return mDrawerLayout.isShowDrawer(gravity);
    }

    @Override
    public int getMenuFrameId() {
        return mMenuView.getId();
    }

    @Override
    public ViewGroup getRootView() {
        return mDrawerLayout;
    }

    @Override
    public ViewGroup getMenuView() {
        return mMenuView;
    }

    @Override
    public ViewGroup getContentView() {
        return mContentView;
    }

    @Override
    public void setContentView(View v) {
        mContentView.addView(v);
    }

    @Override
    public void showMenu(boolean show) {
        if (show) {
            mMenuView.setVisibility(View.VISIBLE);
        } else {
            mMenuView.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean isShowMenu() {
        return mMenuView.getVisibility() == View.VISIBLE;
    }

    @Override
    public void setMenuEnable(boolean enable) {
        if (enable) {
            mMenuView.setVisibility(View.VISIBLE);
        } else {
            mMenuView.setVisibility(View.GONE);
        }
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (mDrawerLayout.isShowDrawer(Gravity.LEFT)) {
                mDrawerLayout.showDrawer(Gravity.LEFT, false);
                return true;
            } else if (mDrawerLayout.isShowDrawer(Gravity.RIGHT)) {
                mDrawerLayout.showDrawer(Gravity.RIGHT, false);
                return true;
            }
        }
        return false;
    }

    @Override
    public void attachToActivity(Activity activity) {
        mDrawerLayout.attachToActivity(activity, ((TabDrawerNavigationFragmentActivity) activity).isFloatActionBarEnabled());
    }

    @Override
    public BaseNavigationFragmentActivity getActivity() {
        return mActivity;
    }

    @Override
    public void setBackgroundColor(int color) {
        mDrawerLayout.setBackgroundColor(color);
    }

    @Override
    public void setBackgroundResource(int resId) {
        mDrawerLayout.setBackgroundResource(resId);
    }

    public void setDrawer(int gravity, BaseFragment fragment) {
        fragment.setUserVisibleHint(false);
        FragmentTransaction t = mActivity.getSupportFragmentManager()
                .beginTransaction();
        t.replace(gravity == Gravity.LEFT ? R.id.left_view : R.id.right_view, fragment, "DrawerFragment_" + gravity);
        t.commitAllowingStateLoss();
        mActivity.getSupportFragmentManager().executePendingTransactions();
    }

    public BaseFragment getDrawer(int gravity) {
        Fragment fragment = mActivity.getSupportFragmentManager().findFragmentById(gravity == Gravity.LEFT ? R.id.left_view : R.id.right_view);
        if(fragment!=null){
            return (BaseFragment) fragment;
        }else{
            return null;
        }
    }

    public void removeDrawer(int gravity) {
        Fragment fragment = mActivity.getSupportFragmentManager().findFragmentById(gravity == Gravity.LEFT ? R.id.left_view : R.id.right_view);
        if(fragment!=null){
            FragmentTransaction t = mActivity.getSupportFragmentManager()
                    .beginTransaction();
            t.remove(fragment);
            t.commitAllowingStateLoss();
            mActivity.getSupportFragmentManager().executePendingTransactions();
        }
    }

}

File: uiframe/src/main/kotlin/mobi/cangol/mobile/navigation/TabDrawerNavigationFragmentActivity.kt
package mobi.cangol.mobile.navigation

import android.app.Activity
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.widget.DrawerLayout
import android.view.*
import android.widget.FrameLayout
import mobi.cangol.mobile.base.BaseFragment
import mobi.cangol.mobile.base.BaseNavigationFragmentActivity

import mobi.cangol.mobile.uiframe.R

/**
 * Created by xuewu.wei on 2017/9/19.
 */

abstract class TabDrawerNavigationFragmentActivity : BaseNavigationFragmentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        this.setNavigationFragmentActivityDelegate(TabDrawerNavigationFragmentActivityDelegate(
                this))
        super.onCreate(savedInstanceState)
        this.customActionBar.titleGravity = Gravity.CENTER
        this.customActionBar.setDisplayShowHomeEnabled(false)
        //默认是不能滑动的
        this.setDrawerEnable(Gravity.LEFT, false)
        this.setDrawerEnable(Gravity.RIGHT, false)
    }

    override fun onSupportNavigateUp(): Boolean {
        return if (stack!!.size() <= 1) {
            stack!!.peek()!!.onSupportNavigateUp()
            true
        } else {
            super.onSupportNavigateUp()
        }
    }

    fun setDrawerEnable(gravity: Int, enable: Boolean) {
        if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).setDrawerEnable(gravity, enable)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    fun showDrawer(gravity: Int, show: Boolean) {
        if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).showDrawer(gravity, show)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    fun isShowDrawer(gravity: Int): Boolean {
        return if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).isShowDrawer(gravity)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    fun setDrawer(gravity: Int, fragmentClass: Class<out BaseFragment>, args: Bundle?) {
        val drawerFragment = Fragment.instantiate(this, fragmentClass.name, args) as BaseFragment
        if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).setDrawer(gravity, drawerFragment)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    fun getDrawer(gravity: Int): BaseFragment? {
        return if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).getDrawer(gravity)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    fun removeDrawer(gravity: Int) {
        if (getNavigationFragmentActivityDelegate() is TabDrawerNavigationFragmentActivityDelegate) {
            (getNavigationFragmentActivityDelegate() as TabDrawerNavigationFragmentActivityDelegate).removeDrawer(gravity)
        } else {
            throw IllegalStateException(GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE)
        }
    }

    companion object {
        const val GET_ACTIVITY_IS_TAB_DRAWER_NAVIGATION_FRAGMENT_ACTIVITY_DELEGATE = "getActivity is TabDrawerNavigationFragmentActivityDelegate"
    }
}

internal class TabDrawerNavigationFragmentActivityDelegate(
        private val mActivity: BaseNavigationFragmentActivity) : AbstractNavigationFragmentActivityDelegate() {
    private var mDrawerLayout: TabMenuDrawerLayout? = null
    private var mMenuView: FrameLayout? = null
    private var mContentView: FrameLayout? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        mDrawerLayout = LayoutInflater.from(mActivity).inflate(
                R.layout.navigation_tab_drawer_main, null) as TabMenuDrawerLayout

        mContentView = mDrawerLayout!!.findViewById(R.id.content_view)
        mMenuView = mDrawerLayout!!.findViewById(R.id.menu_view)
        mDrawerLayout?.setDrawerListener(object : DrawerLayout.DrawerListener {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
                val fragment = mActivity.supportFragmentManager.findFragmentById(drawerView.id)
                if (fragment is BaseFragment) {
                    fragment.onDrawerSlide(slideOffset)
                }
            }

            override fun onDrawerOpened(drawerView: View) {
                val fragment = mActivity.supportFragmentManager.findFragmentById(drawerView.id)
                if (fragment is BaseFragment) {
                    fragment.userVisibleHint = true
                    fragment.onDrawerOpened()
                }
            }

            override fun onDrawerClosed(drawerView: View) {
                val fragment = mActivity.supportFragmentManager.findFragmentById(drawerView.id)
                if (fragment is BaseFragment) {
                    fragment.userVisibleHint = false
                    fragment.onDrawerClosed()
                }
            }

            override fun onDrawerStateChanged(newState: Int) {
                //do nothings
            }
        })
    }

    fun setDrawerEnable(gravity: Int, enable: Boolean) {
        mDrawerLayout!!.setDrawerEnable(gravity, enable)
    }

    override fun getMaskView(): FrameLayout {
        return mDrawerLayout!!.getMaskView()
    }

    override fun displayMaskView(show: Boolean) {
        mDrawerLayout!!.displayMaskView(show)
    }

    fun showDrawer(gravity: Int, show: Boolean) {
        mDrawerLayout!!.showDrawer(gravity, show)
    }

    fun isShowDrawer(gravity: Int): Boolean {
        return mDrawerLayout!!.isShowDrawer(gravity)
    }

    override fun getMenuFrameId(): Int {
        return mMenuView!!.id
    }

    override fun getRootView(): ViewGroup {
        return mDrawerLayout!!
    }

    override fun getMenuView(): ViewGroup {
        return mMenuView!!
    }

    override fun getContentView(): ViewGroup {
        return mContentView!!
    }

    override fun setContentView(v: View) {
        mContentView!!.addView(v)
    }

    override fun showMenu(show: Boolean) {
        if (show) {
            mMenuView!!.visibility = View.VISIBLE
        } else {
            mMenuView!!.visibility = View.GONE
        }
    }

    override fun isShowMenu(): Boolean {
        return mMenuView!!.visibility == View.VISIBLE
    }

    override fun setMenuEnable(enable: Boolean) {
        if (enable) {
            mMenuView!!.visibility = View.VISIBLE
        } else {
            mMenuView!!.visibility = View.GONE
        }
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (mDrawerLayout!!.isShowDrawer(Gravity.LEFT)) {
                mDrawerLayout!!.showDrawer(Gravity.LEFT, false)
                return true
            } else if (mDrawerLayout!!.isShowDrawer(Gravity.RIGHT)) {
                mDrawerLayout!!.showDrawer(Gravity.RIGHT, false)
                return true
            }
        }
        return false
    }

    override fun attachToActivity(activity: Activity) {
        mDrawerLayout!!.attachToActivity(activity, (activity as TabDrawerNavigationFragmentActivity).isFloatActionBarEnabled())
    }

    override fun getActivity(): BaseNavigationFragmentActivity {
        return mActivity
    }

    override fun setBackgroundColor(color: Int) {
        mDrawerLayout!!.setBackgroundColor(color)
    }

    override fun setBackgroundResource(resId: Int) {
        mDrawerLayout!!.setBackgroundResource(resId)
    }

    fun setDrawer(gravity: Int, fragment: BaseFragment) {
        fragment.userVisibleHint = false
        val t = mActivity.supportFragmentManager
                .beginTransaction()
        t.replace(if (gravity == Gravity.LEFT) R.id.left_view else R.id.right_view, fragment, "DrawerFragment_$gravity")
        t.commitAllowingStateLoss()
        mActivity.supportFragmentManager.executePendingTransactions()
    }

    fun getDrawer(gravity: Int): BaseFragment? {
        val fragment = mActivity.supportFragmentManager.findFragmentById(if (gravity == Gravity.LEFT) R.id.left_view else R.id.right_view)
        return if (fragment != null) {
            fragment as BaseFragment?
        } else {
            null
        }
    }

    fun removeDrawer(gravity: Int) {
        val fragment = mActivity.supportFragmentManager.findFragmentById(if (gravity == Gravity.LEFT) R.id.left_view else R.id.right_view)
        if (fragment != null) {
            val t = mActivity.supportFragmentManager
                    .beginTransaction()
            t.remove(fragment)
            t.commitAllowingStateLoss()
            mActivity.supportFragmentManager.executePendingTransactions()
        }
    }

}
